### PR TITLE
Spreadsheet: Add undo/redo functionality

### LINKED ISSUE #4270: Spreadsheet: Add support for undo/redo
Should help the unlucky people who managed to copy things into the wrong place unintentionally lol

### PR DESCRIPTION
**Pull request by hikilaka**
Giving a little love to Spreadsheet in this PR.

This PR adds functionality for:
 - Undo/Redo (this would close #4270).
 - Add window()->set-modified if a change occurs.

### CODE REVIEW

File: Userland/Applications/Spreadsheet/Cell.cpp [hunk 1]
```diff
@@ -191,19 +191,21 @@ void Cell::copy_from(const Cell& other)
     m_thrown_value = other.m_thrown_value;
 }
 
-CellUndoCommand::CellUndoCommand(Cell& cell)
+CellUndoCommand::CellUndoCommand(Cell& cell, String& previous_data)
```
Comment (alimpfard) [inline Userland/Applications/Spreadsheet/Cell.cpp:194]:
```suggestion
CellUndoCommand::CellUndoCommand(Cell& cell, String const& previous_data)
```

**Review by alimpfard — COMMENTED** [review]
This seems like too many commits that just modify the previous one, perhaps these could be better factored, or just squashed?

File: Userland/Applications/Spreadsheet/Cell.cpp (hunk 1 above)
Comment (hikilaka) [reply Userland/Applications/Spreadsheet/Cell.cpp:194]: Thanks for the feedback! I can agree. My development process isn't really pertinent to this feature.

... as for the `const` addition, I'll add that :-) +1 for const-correctness!

File: Userland/Applications/Spreadsheet/SpreadsheetWidget.cpp [hunk 2]
```diff
@@ -227,6 +235,8 @@ SpreadsheetWidget::SpreadsheetWidget(GUI::Window& parent_window, NonnullRefPtrVe
     toolbar.add_action(*m_cut_action);
     toolbar.add_action(*m_copy_action);
     toolbar.add_action(*m_paste_action);
+    toolbar.add_action(*m_undo_action);
+    toolbar.add_action(*m_redo_action);
```
[the remark this answers is not in the document]
Comment (hikilaka) [reply Userland/Applications/Spreadsheet/SpreadsheetWidget.cpp:239]: I had the thought of doing so, looked at PixelPaint, saw they didn't do it this way, and went on my way :-) 

I do like the aesthetic of your suggestions, though!

File: Userland/Applications/Spreadsheet/SpreadsheetWidget.cpp [hunk 3]
```diff
@@ -537,6 +571,8 @@ void SpreadsheetWidget::initialize_menubar(GUI::Window& window)
     edit_menu.add_action(*m_cut_action);
     edit_menu.add_action(*m_copy_action);
     edit_menu.add_action(*m_paste_action);
+    edit_menu.add_action(*m_undo_action);
+    edit_menu.add_action(*m_redo_action);
```
[the remark this answers is not in the document]
Comment (hikilaka) [reply Userland/Applications/Spreadsheet/SpreadsheetWidget.cpp:577]: see above :-)

**Comment by alimpfard** [comment]
Looks alright to me, let's just wait for CI.

**Comment by hikilaka** [comment]
TFW you accidentally delete a branch that is awaiting to be PR'd 😅

**Comment by alimpfard** [comment]
Alright, let's get this in.

**Pull request merged**